queue/queueservice.js

```diff
@@ -303,15 +315,140 @@ class QueueService {
     ].join('-');
 
     // Return and cache promise that we created this queue
-    return this.queues[id] = this._createQueue(name).catch(err => {
+    return this.queues[id] = Promise.all([
+      this._ensureQueueAndMetadata(name, provisionerId, workerType),
+      this._ensureQueueAndMetadata(legacyName, provisionerId, workerType)
+    ]).catch(err => {
+      debug("[alert-operator] Failed to ensure azure queue: %s, as JSON: %j",
+            err, err, err.stack);
+
       // Don't cache negative results
       this.queues[id] = undefined;
       throw err;
-    }).then(() => {
-      return name;
-    });
+    }).then(() => name);
+  }
+
+  /**
+   * Ensure that queue with given name exists and has correct meta-data,
+   * in particular regarding meta-data for expiration.
+   *
+   * We maintain following meta-data keys:
+   *  * `provisioner_id`,
+   *  * `worker_type`, and
+   *  * `last_used` (tolerating 23 hours difference, updated every 25 hours)
+   *
+   * We delete queues if they have `last_used` > 10 days ago, this happens in
+   * a periodic test tasks.
+   */
+  async _ensureQueueAndMetadata(queue, provisionerId, workerType) {
+    // Fetch meta-data from queue, checking if it exists
+    try {
+      let {metadata} = await this.client.getMetadata(queue);
+
+      // Check if meta-data is up-to-date
+      var lastUsed = new Date(metadata.last_used);
+      if (metadata.provisioner_id === provisionerId
+          && metadata.worker_type === workerType
+          && isFinite(lastUsed)
+          && lastUsed.getTime() > Date.now() - 23 * 60 * 60 * 1000
+      ) {
+        return; // We're done as meta-data is present
+      }
+
+      // Update meta-data
+      return this.client.setMetadata(queue, {
+        provisioner_id: provisionerId,
+        worker_type:    workerType,
+        last_used:      taskcluster.fromNowJSON()
+      });
+    } catch (err) {
+      // We handle queue not found exceptions, because getMetadata is a HEA
+      // request we don't get any error message payload, so we also accept 404
+      // as implying the same.
+      if (err.code !== 'QueueNotFound' &&
+          err.statusCode !== 404) {
+        throw err;
+      }
+
+      // Create the queue with correct meta-data
+      try {
+        await this.client.createQueue(queue, {
+          provisioner_id: provisionerId,
+          worker_type:    workerType,
+          last_used:      taskcluster.fromNowJSON()
+        });
+      } catch (err) {
+        // If queue already exists, we must have been racing we assume meta-data
+        // is up to date...
+        if (err.code !== 'QueueAlreadyExists') {
+          // We probably don't have to alert on all these. But we should
+          // definitely alert if we see a QueueBeingDeleted error
+          debug("[alert-operator] Failed to create queue: %s for %s/%s " +
```

Comment:
Oh, so instead of having the `+` ?

Doing `debug('[alert-operator] Failed to create queue: ${queue} '...)` is a lot less efficient than `debug('[alert-operator] Failed to create queue: %s '..., queue)`, because `debug` doesn't do the string formatting if `DEBUG` isn't set... Then the `debug` function is just an empty function if the particular file has been disabled, hence, no overhead.